clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -519,26 +550,52 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
-            // XXX: Handle the back-to-back signals case where the second
-            // signal arrives just after the return from the first without
-            // any intervening instrs. The return point of the second one
-            // would be the pc in the kernel xfer marker of the first.
-            shard->prev_xfer_marker_.marker.marker_type !=
-                TRACE_MARKER_TYPE_KERNEL_XFER) {
-            if (shard->saw_kernel_xfer_after_prev_instr_) {
-                // We have nested signals without an intervening app instr.
-                // Push an empty instr to mean that this shouldn't be used.
-                shard->pre_signal_instr_.push({});
+        if (memref.marker.marker_type == TRACE_MARKER_TYPE_KERNEL_XFER) {
+            // We assume paired signal entry-exit (so no longjmp and no rseq
+            // inside signal handlers).
+            if (shard->prev_xfer_int_pc_.empty() ||
+                shard->pre_signal_instr_.size() == 1 ||
+                shard->prev_xfer_aborted_rseq_.empty()) {
+                // If one is true, all should be so.
+                assert(shard->prev_xfer_int_pc_.empty() &&
+                       shard->pre_signal_instr_.size() == 1 &&
+                       shard->prev_xfer_aborted_rseq_.empty());
+                // This can happen if tracing started in the middle of a signal.
+                // Try to continue by skipping the checks.
+                shard->last_xfer_int_pc_ = 0;
+                shard->last_pre_signal_instr_ = {};
+                shard->last_xfer_aborted_rseq_ = false;
+                // Now that we know there's an outer context that we didn't see before,
+                // forget the last instr we saw in what was incorrectly assumed previously
+                // to be the outer-most context.
+                replace_top<memref_t>(shard->pre_signal_instr_, {});
+            } else {
+                shard->last_xfer_int_pc_ = shard->prev_xfer_int_pc_.top();
+                shard->prev_xfer_int_pc_.pop();
+                shard->pre_signal_instr_.pop();
+                // The last instr in the context immediately outer to this signal may
+                // be {} in some cases:
+                // - for nested signals without any intervening instr
+                // - if there's a signal at the very beginning of the trace
+                // In both these cases the empty instr implies that it should not
+                // be used for the pre-signal instr check.
+                shard->last_pre_signal_instr_ = shard->pre_signal_instr_.top();
+                shard->last_xfer_aborted_rseq_ = shard->prev_xfer_aborted_rseq_.top();
+                shard->prev_xfer_aborted_rseq_.pop();
+            }
+        }
+        if (memref.marker.marker_type == TRACE_MARKER_TYPE_KERNEL_EVENT) {
+            // If preceded by an RSEQ abort marker, this is not really a signal.
+            if (shard->prev_entry_.marker.marker_type == TRACE_MARKER_TYPE_RSEQ_ABORT) {
```

Comment:
Doesn't this need to first check that prev_entry_.marker.type is a marker?

Reply:
That's right. I'll add that.

Reply:
Done.